src/extension/serializer.ts

```diff
@@ -733,26 +735,41 @@ export class GrpcSerializer extends SerializerBase {
     return await workspace.applyEdit(edit)
   }
 
+  protected notebookDataToProto(data: NotebookData): typeof this.protoNotebookType {
+    return Notebook.clone(data as any)
+  }
+
+  getSerializationRequest(obj: Object): SerializeRequest {
+    return <SerializeRequest>{ ...obj }
+  }
+
   protected async saveNotebook(
     data: NotebookData,
     // eslint-disable-next-line @typescript-eslint/no-unused-vars
     token: CancellationToken,
   ): Promise<Uint8Array> {
     const marshalFrontmatter = this.lifecycleIdentity === RunmeIdentity.ALL
 
-    const notebook = GrpcSerializer.marshalNotebook(data, { marshalFrontmatter })
+    const notebookProto = this.notebookDataToProto(data)
+    const notebook = Marshal.notebook(data, notebookProto, {
+      marshalFrontmatter,
+    }) as typeof this.protoNotebookType
 
     if (marshalFrontmatter) {
       data.metadata ??= {}
       data.metadata[RUNME_FRONTMATTER_PARSED] = notebook.frontmatter
     }
 
-    const cacheId = GrpcSerializer.getDocumentCacheId(data.metadata)
+    const cacheId = SerializerBase.getDocumentCacheId(data.metadata)
     this.notebookDataCache.set(cacheId as string, data)
 
-    const serialRequest = <SerializeRequest>{ notebook }
+    const serialRequest = this.getSerializationRequest({ notebook })
 
-    const cacheOutputs = this.cacheNotebookOutputs(notebook, cacheId)
+    const maskedNotebookProto = this.notebookDataToProto(data)
+    const maskedNotebook = Marshal.notebook(data, maskedNotebookProto, {
+      marshalFrontmatter,
+    }) as typeof this.protoNotebookType
+    const cacheOutputs = this.cacheNotebookOutputs(notebook, maskedNotebook, cacheId)
```

Comment:
How come the signature of this.cacheNotebookOutputs changed?

Reply:
some background:
upstream `marshalNotebook` augments and converts a vscode notebook to a protobuf-ts `Notebook`. I wanted to have it be proto agnostic for reuse in grpc and connect serializers.  [ This ](https://github.com/hotpocket/vscode-runme/pull/2/commits/1a00cdb354574b684088479e0e4ed83333e41d59) was my the start of my attempt, which created a `Marshal` class allowing abstraction via composition.  Unfortunately I underestimated the level of integration of the operations of the class and the proto types.  Since then, I have started to store type information directly in the class via [properties](https://github.com/hotpocket/vscode-runme/commit/7c963414508b296a1ea4b01f2d2bef59c1acb977#diff-f28bac02639ff6caf3f3123caf29961e753eb6b74637f5d893a51d935d49f162R510) and [methods](https://github.com/hotpocket/vscode-runme/commit/7c963414508b296a1ea4b01f2d2bef59c1acb977#diff-f28bac02639ff6caf3f3123caf29961e753eb6b74637f5d893a51d935d49f162R749). 

the method signature changed due to my original efforts at abstraction via composition.  Since then, given the class properties & methods, I could likely re-integrate this and move the method signature back to the original(this is on the todo list).  I need to clean this up for public consumption.